### PR TITLE
Feature kernel semaphore syscalls

### LINKED ISSUE #389: [pm] Implement Semaphore Kernel Calls
## Description
Add syscalls in libnanvix and kernel: semget, semop and semctl.
Create test functions.

### semget
`int kcall_semop (int id, int op)`
If key already exist, associate current process at semaphore, else, create a semaphore with the key.

### semop
`int kcall_semop (int id, int op)`
Verify if process can operate in semaphore (id), and operate (op):

### semctl
`int kcall_semctl (int id, int cmd, int val)`
Verify if process can control semaphore (id), and exec (cmd):

### Change:
* `include/nanvix/kernel/kcall.h`
*  `src/kernel/kcall/mod.c`
*  `src/kernel/kcall/mod.h`
* `src/libnanvix/src/kcall/mod.rs`
* `src/libnanvix/src/pm/mod.rs`
* `src/servers/test/src/test.rs`

### Add:
* `src/kernel/kcall/sem/semget.c`
* `src/kernel/kcall/sem/semop.c`
* `src/kernel/kcall/sem/semctl.c`

Root issue: #385

### PR DESCRIPTION
**Pull request by rafael0121**
## Description

- This PR closes #389

## Summary of Changes

- Implemented *Semaphore Kernel Calls*, and added test functions in server *test* .

### CODE REVIEW

File: src/kernel/kcall/sem/semctl.c [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#include <nanvix/kernel/lib.h>
+#include <nanvix/kernel/pm/process.h>
+#include <nanvix/kernel/pm/semaphore.h>
+
+int kcall_semctl(int id, int cmd, int val)
```
Comment (ppenna) [inline src/kernel/kcall/sem/semctl.c:5]: Add a brief comment on what this function does here. Every function should have a comment. You can copy the "brief" from the public doc. This helps new commers to understand the code without searching for public docs.

File: src/kernel/kcall/sem/semctl.c [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#include <nanvix/kernel/lib.h>
+#include <nanvix/kernel/pm/process.h>
+#include <nanvix/kernel/pm/semaphore.h>
```
Comment (ppenna) [inline src/kernel/kcall/sem/semctl.c:3]: Enclose this in a section. Check out how other source files do this. This is to maintain consistency across the code base.

File: src/kernel/kcall/sem/semctl.c [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+int kcall_semctl(int id, int cmd, int val)
+{
+    // TODO
+    return id + cmd + val;
```
Comment (ppenna) [inline src/kernel/kcall/sem/semctl.c:8]: Enclose this with `()` to follow existing coding standards.

File: src/kernel/kcall/sem/semget.c [hunk 4]
```diff
@@ -0,0 +1,10 @@
+#include <nanvix/kernel/lib.h>
+#include <nanvix/kernel/pm/process.h>
+#include <nanvix/kernel/pm/semaphore.h>
```
Comment (ppenna) [inline src/kernel/kcall/sem/semget.c:3]: Enclose this in a section. Check out how other source files do this. This is to maintain consistency across the code base.

File: src/kernel/kcall/sem/semget.c [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+
+// Just a kernel call test.
+int kcall_semget(unsigned key)
+{
+    // TODO
+    return key;
```
Comment (ppenna) [inline src/kernel/kcall/sem/semget.c:9]: Enclose this with `()` to follow existing coding standards.

File: src/kernel/kcall/sem/semop.c [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#include <nanvix/kernel/lib.h>
+#include <nanvix/kernel/pm/process.h>
+#include <nanvix/kernel/pm/semaphore.h>
```
Comment (ppenna) [inline src/kernel/kcall/sem/semop.c:3]: Enclose this in a section. Check out how other source files do this. This is to maintain consistency across the code base.

File: src/kernel/kcall/sem/semop.c [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+
+// Just a kernel call test.
+int kcall_semop(int id, int op)
+{
+    // TODO
+    return id + op;
```
Comment (ppenna) [inline src/kernel/kcall/sem/semop.c:9]: Enclose this with `()` to follow existing coding standards.

File: src/kernel/kcall/sem/semget.c [hunk 8]
```diff
@@ -0,0 +1,10 @@
+#include <nanvix/kernel/lib.h>
+#include <nanvix/kernel/pm/process.h>
+#include <nanvix/kernel/pm/semaphore.h>
+
+// Just a kernel call test.
```
Comment (ppenna) [inline src/kernel/kcall/sem/semget.c:5]: Update this to something useful.

File: src/kernel/kcall/sem/semop.c [hunk 9]
```diff
@@ -0,0 +1,10 @@
+#include <nanvix/kernel/lib.h>
+#include <nanvix/kernel/pm/process.h>
+#include <nanvix/kernel/pm/semaphore.h>
+
+// Just a kernel call test.
```
Comment (ppenna) [inline src/kernel/kcall/sem/semop.c:5]: Update this to something useful.

File: src/servers/test/src/test.rs [hunk 10]
```diff
@@ -423,6 +424,41 @@ fn get_invalid_kmod_info() -> bool {
     true
 }
 
+/* Test if Semaphore kernel calls is working. */
```
Comment (ppenna) [inline src/servers/test/src/test.rs:427]: Use `//` for single-line comments.

File: src/servers/test/src/test.rs [hunk 11]
```diff
@@ -423,6 +424,41 @@ fn get_invalid_kmod_info() -> bool {
...
+    let key: u32 = 1;
+    let result: i32 = pm::semget(key);
+    if result != 1 {
+        return false;
+    }
+
+    true
+}
+
+fn test_semop_call() -> bool {
```
Comment (ppenna) [inline src/servers/test/src/test.rs:439]: Add a comment on this function

**Review by ppenna — COMMENTED** [review]
Add a copyright note to all new files.

File: src/kernel/kcall/sem/semctl.c (hunk 2 above)
Comment (rafael0121) [reply src/kernel/kcall/sem/semctl.c:12]: Enclosed.

File: src/kernel/kcall/sem/semctl.c (hunk 1 above)
Comment (rafael0121) [reply src/kernel/kcall/sem/semctl.c:21]: Added.

File: src/kernel/kcall/sem/semget.c (hunk 4 above)
Comment (rafael0121) [reply src/kernel/kcall/sem/semget.c:12]: Enclosed.

File: src/kernel/kcall/sem/semget.c (hunk 5 above)
Comment (rafael0121) [reply src/kernel/kcall/sem/semget.c]: Enclosed.

File: src/kernel/kcall/sem/semop.c (hunk 6 above)
Comment (rafael0121) [reply src/kernel/kcall/sem/semop.c:12]: Enclosed.

File: src/servers/test/src/test.rs [hunk 12]
```diff
@@ -423,6 +424,41 @@ fn get_invalid_kmod_info() -> bool {
...
+    let id: u32 = 1;
+    let op: u32 = 1;
+    let result: i32 = pm::semop(id, op);
+    if result != 2 {
+        return false;
+    }
+
+    true
+}
+fn test_semctl_call() -> bool {
```
[the remark this answers is not in the document]
Comment (rafael0121) [reply src/servers/test/src/test.rs:451]: Added.

File: src/servers/test/src/test.rs [hunk 13]
```diff
@@ -423,6 +424,41 @@ fn get_invalid_kmod_info() -> bool {
...
+fn test_semop_call() -> bool {
+    let id: u32 = 1;
+    let op: u32 = 1;
+    let result: i32 = pm::semop(id, op);
+    if result != 2 {
+        return false;
+    }
+
+    true
+}
```
[the remark this answers is not in the document]
Comment (rafael0121) [reply src/servers/test/src/test.rs:448]: Added.

File: src/servers/test/src/test.rs (hunk 11 above)
Comment (rafael0121) [reply src/servers/test/src/test.rs:439]: Added.

**Review by ppenna — APPROVED** [review]
Great! Thanks for working on this. Once regression is fixed please squash all commits into one.

**Pull request merged**